docs/source/type_narrowing.rst

```diff
@@ -0,0 +1,289 @@
+.. _type-narrowing:
+
+Type narrowing
+==============
+
+This section is dedicated to  several type narrowing
+techniques which are supported by Mypy.
+
+What is type narrowing?
+It is an operation of telling type checking that some
+broader type is actually more specific.
+
+
+Type narrowing expressions
+--------------------------
+
+The simplest way to narrow a type is to use one of the supported expressions:
+
+- :py:func:`type` like in ``type(obj) is int`` will narrow ``obj`` to have ``int`` type
+- :py:func:`isinstance` like in ``isinstance(obj, float)`` will narrow ``obj`` to have ``float`` type
+- :py:func:`issubclass` like in ``issubclass(cls, MyClass)`` will narrow ``cls`` to be ``Type[MyClass]``
+
+It is important to understand that type narrowing happens in a specific context.
+It can be a single ``if`` branch:
+
+.. code-block:: python
+
+  def function(arg: object):
+      if isinstance(arg, int):
+          # Type is narrowed for this ``if`` branch context only
+          reveal_type(arg)  # Revealed type: "builtins.int"
+      elif isinstance(arg, str) or isinstance(arg, bool):
+          # This ``if`` branch has new context for narrowing:
+          reveal_type(arg)  # Revealed type: "builtins.str | builtins.bool"
+
+          # Contexts can be nested:
+          if isinstance(arg, bool):
+              reveal_type(arg)  # Revealed type: "builtins.bool"
+
+      # In the main context type won't be changed:
+      reveal_type(arg)  # Revealed type: "builtins.object"
+
+We can also use reverse logic as well:
+
+.. code-block:: python
+
+  def function(arg: int | str):
+      if isinstance(arg, int):
+          return
+
+      # `arg` can't be `int` at this point:
+      reveal_type(arg)  # Revealed type: "builtins.str"
+
+We can also use ``assert`` to narrow types in the same context:
+
+.. code-block:: python
+
+  def function(arg: Any):
+      assert isinstance(arg, int)
+      reveal_type(arg)  # Revealed type: "builtins.int"
+
+Mypy can also use :py:func:`issubclass`
+for better type inference when working with types and metaclasses:
+
+.. code-block:: python
+
+   class MyCalcMeta(type):
+       @classmethod
+       def calc(cls) -> int:
+           ...
+
+   def f(o: object) -> None:
+       t = type(o)  # We must use a variable here
+       reveal_type(t)  # Revealed type is "builtins.type"
+
+       if issubtype(t, MyCalcMeta):  # `issubtype(type(o), MyCalcMeta)` won't work
+           reveal_type(t)  # Revealed type is "Type[MyCalcMeta]"
+           t.calc()  # Okay
+
+.. note::
+
+  Narrowing type for some impossible state will be treated as an error.
```

Comment:
This isn't generally true, e.g. not even with `--strict`, I think you need `--warn-unreachable` for this.

Reply:
Could also maybe mention the `assert_never` trick here though